File: BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#Time COmplexity: O(1) on average
#Sapce Complexiy : O(H) where H is the height of tree
class BSTIterator(object):
    st = []
    def dfs(self,root) :
        while(root!=None):
            self.st.append(root)
            root =root.left

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.dfs(root)

    def next(self):
        """
        @return the next smallest number
        :rtype: int
        """
        node =self.st.pop()
        self.dfs(node.right)
        return node.val

    def hasNext(self):
        """
        @return whether we have a next smallest number
        :rtype: bool
        """
        if len(self.st)!= 0:
            return True
        
    
    


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: bst.java

// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no

import java.util.*;
// Your code here along with comments explaining your approach
class TreeNode {
       int val;
         TreeNode left;
         TreeNode right;
         TreeNode() {}
         TreeNode(int val) { this.val = val; }
         TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
 }

  public class bst {
       Stack<TreeNode>st;
       public bst(TreeNode root) {
           st=new Stack<>();
           dfs(root);
       }
       
       public int next() {
           TreeNode result=st.pop();
           dfs(result.right);
           return result.val;
       }
       
       public boolean hasNext() {
           return !st.isEmpty();
       }
       private void dfs(TreeNode root){
           while(root!=null){
               st.push(root);
               root=root.left;
           }
       }
   }
